### PR TITLE
storage: Add metrics for tracking CommandQueue health

### PR DESCRIPTION
**Pull request by nvb**
Related to #9797.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10067)

### CODE REVIEW

File: pkg/storage/command_queue.go [hunk 1]
```diff
@@ -73,6 +81,13 @@ func (c *cmd) Range() interval.Range {
 	return c.key
 }
 
+func (c *cmd) cmdCount() int {
```
Comment (bdarnell) [inline pkg/storage/command_queue.go]: Why isn't this `return len(c.children)+1`? (add a comment)

File: pkg/storage/command_queue.go [hunk 2]
```diff
@@ -52,6 +52,14 @@ type CommandQueue struct {
 	wRg, rwRg interval.RangeGroup // avoids allocating in GetWait
 	oHeap     overlapHeap         // avoids allocating in GetWait
 	overlaps  []*cmd              // avoids allocating in getOverlaps
+
+	// Used to temporarily store metrics local to a single CommandQueue. These
+	// will periodically be processed by the Store.
+	localMetrics struct {
+		readCommands    int64
+		writeCommands   int64
+		maxOverlapsSeen int64
```
Comment (bdarnell) [inline pkg/storage/command_queue.go]: Add a comment that `maxOverlapsSeen` is modified by the store, not just read.

**Review by bdarnell — APPROVED** [review]

**Comment by tschottdorf** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/metrics.go, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10067#-KURPQF0nPW23SjGsGuK:-KURPQF0nPW23SjGsGuL:b-eziktv) ([raw file](https://github.com/cockroachdb/cockroach/blob/edd6f981a13b783f289bf496070cb6fa4651a9e9/pkg/storage/metrics.go#L42)):*
> ```Go
> 
> 	// Replica CommandQueue metrics.
> 	metaMaxCommandQueueSize = metric.Metadata{Name: "replicas.commandqueue.maxsize",
> ```

tracking maximum values is somewhat awkward from a metrics perspective because there's always an implicit second parameter (how long are you taking the max for).

The right way to track this is a histogram or summary, but that seems like more work than appropriate right now. Perhaps just add a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10067#-:-KURP9rmWsVWlgsZnYA3:bnfp4nl)*

**Comment by nvanbenschoten** [comment]
TFTRs.

---

Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/storage/command_queue.go, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10067#-KUPlM_isroizAiZ7YiZ-r1-61:-KUcdfL30koFIk64M-u0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/edd6f981a13b783f289bf496070cb6fa4651a9e9/pkg/storage/command_queue.go#L61)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Add a comment that `maxOverlapsSeen` is modified by the store, not just read.
</details>
Done.

---

*[pkg/storage/command_queue.go, line 84 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10067#-KUPlM_isroizAiZ7YiZ-r1-84:-KUcdg9pXXgH2VxxA94-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/edd6f981a13b783f289bf496070cb6fa4651a9e9/pkg/storage/command_queue.go#L84)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why isn't this `return len(c.children)+1`? (add a comment)
</details>
Done.

---

*[pkg/storage/metrics.go, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10067#-KURPQF0nPW23SjGsGuK:-KUcdgk5IMPPI1d-xjO6:b-m568i0) ([raw file](https://github.com/cockroachdb/cockroach/blob/edd6f981a13b783f289bf496070cb6fa4651a9e9/pkg/storage/metrics.go#L42)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> tracking maximum values is somewhat awkward from a metrics perspective because there's always an implicit second parameter (how long are you taking the max for).
> 
> The right way to track this is a histogram or summary, but that seems like more work than appropriate right now. Perhaps just add a comment.
</details>
Done. And is the maximum size dependent on sampling period? A sum certainly would be, but this is the maximum value of a set of point-in-time measurements. That's different than the maximum value seen over a sampling period (which coincidentally is what `metaMaxCommandQueueOverlaps` is measuring).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10067#-:-KUcdzP8XW77j-2svIUi:bz5maj4)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10067)*

**Pull request merged**